### ISSUE TITLE
RAND_bytes() not supported?

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/libp11#430]
### Config
MacOS Big Sur 11.6.1, Xcode-13.1 (probably both irrelevant), current master.

### Problem
Engine apparently claims to provide `RAND_bytes()` to OpenSSL, but fails to do so. Symptoms:
```
$ openssl rand -engine pkcs11 -hex 8
Engine "pkcs11" set.
$ openssl rand -hex 8
bf26ccfae60a98bb
$
```

This problem also prevents enabling PKCS#11 engine in `openssl.cnf` by default, forcing something like
```
.  .  .  .  .
[engine_section]
#pkcs11 = pkcs11_section
gost = gost_section

[pkcs11_section]
engine_id = pkcs11
dynamic_path = /opt/local/libexec/openssl11/lib/engines-1.1/pkcs11.dylib
MODULE_PATH = /Library/OpenSC/lib/opensc-pkcs11.so
init = 0
.  .  .  .  .
```
If `pkcs11 = pkcs11_section` is not commented out, `openssl rand xxxx` does not work (see https://github.com/openssl/openssl/issues/16996).

### Solution?
IMHO, need to add support for requesting random bytes from the token. I could take a shot at it, if somebody (@mtrojnar ? @dengert ?) would guide me wrt. what/where to add.

### Update
It looks like the fault is in my fork, as the master does not contain the offending line(s):
https://github.com/OpenSC/libp11/blob/af2bca28a06617b5a9742701fd7b5e89816abeec/src/eng_front.c#L213-L240

 still think that it would be nice to enable access to randomness generated by the token, via PKCS#11 engine.

**Comment by mtrojnar** [comment]
While I agree that an alternative/optional entropy source may be useful for some deployments, using PKCS#11 as entropy source by default is way more likely to cause problems.  Physical machines tend to have fairly decent entropy sources available, and virtual machines can rarely use physical PKCS#11 tokens.

**Comment by mouse07410** [comment]
>  Physical machines tend to have fairly decent entropy sources available  .  .  .

I agree.

> .  .  .  and virtual machines can rarely use physical PKCS#11 tokens

I agree.

> .  .  .  using PKCS#11 as entropy source by default is way more likely to cause problems

I'm not sure what you mean. 

I'd like to see a way for an app to be able to obtain entropy from a hardware token via PKCS#11. I do not claim that it should be the default. But, as I understand, currently there's no way for, e.g., OpenSSL to do `openssl rand -engine pkcs11 NNN`. I'd like to see this feature added.

**Comment by mtrojnar** [comment]
> > .  .  .  using PKCS#11 as entropy source by default is way more likely to cause problems
> 
> I'm not sure what you mean. 

Precisely the problems described in https://github.com/openssl/openssl/issues/16996

> I'd like to see a way for an app to be able to obtain entropy from a hardware token via PKCS#11. I do not claim that it should be the default. But, as I understand, currently there's no way for, e.g., OpenSSL to do `openssl rand -engine pkcs11 NNN`. I'd like to see this feature added.

A more detailed rationale might help prioritize this requirement.

**Comment by mouse07410** [comment]
> > >    . . . using PKCS#11 as entropy source by default is way more likely to cause problems
> >
> >   I'm not sure what you mean.
>
> Precisely the problems described in openssl/openssl#16996

Respectfully disagree - absence of support of PKCS#11 as entropy source was _precisely_ the cause of openssl/openssl#16996. The engine claimed to provide it, but it was a no-op.  So, I'm "lobbying" for replacing "no-op" with a real implementation.

> A more detailed rationale might help prioritize this requirement.

I have an HSM that's accessible via PKCS#11 (by the "guard" machine, but that detail is irrelevant here). Apps that use HSM services, want to get entropy from it. Some - want it instead of the CPU- or OS-provided entropy. Others (the smarter ones ;) - want to augment the entropy obtained from CPU (RDSEED/RDRAND) with entropy obtained from the HSM, to ensure highest quality. 

Same applies to smart cards - being able to augment whatever entropy you have with off-board hardware-produced bits is helpful, from cryptography point of view.

**Comment by mouse07410** [comment]
@mtrojnar do you concur with my rationale?

**Comment by mouse07410** [comment]
@mtrojnar several months later - I think having the ability to retrieve randomness from a hardware source not related to the computer CPU is a good thing. After all, nobody is forced to use this source of randomness against his will.

How can we make `C_GenerateRandom()` available from `libp11` via PKCS#11?
